src/shared/clients/defaultCloudWatchLogsClient.ts

```diff
@@ -0,0 +1,39 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { CloudWatchLogs } from 'aws-sdk'
+import { ext } from '../extensionGlobals'
+import { CloudWatchLogsClient } from './cloudWatchLogsClient'
+
+export class DefaultCloudWatchLogsClient implements CloudWatchLogsClient {
+    public constructor(public readonly regionCode: string) {}
+
+    public async *describeLogGroups(): AsyncIterableIterator<CloudWatchLogs.LogGroup> {
```

Comment:
I saw this approach in another impl. I don't think what it's doing here is actually desired behavior.

Consider the usage of this. It loads the entire next page on demand as soon as it's needed. 

On the one hand, the caller is completely unaware of this service call. This makes it easier for them to use this.

On the other hand, the caller *should* be aware of the service call. The normal use case would be to coordinate the pagination with a "Load more..." action. The caller is then completely aware of exactly what they have and exactly when a service call is made.

Basically in order to use an async iterable here, you would return an async interable of *pages of log groups* rather than an async interable of *log groups*.

That gives the caller control over pagination without them needing to keep track of the token.
That being said, I honestly think the caller is better off just knowing about the continuation token. It avoids the caller unintentionally treating iteration as a "cheap" operation when it actually results in a service call.

Reply:
I think I'll hold off on this change for now. Ultimately, unless we want a "Load More..." button, the caller is going to essentially implement the same do-while loop, which I guess makes the caller more conscious but at the same time we're pretty much using these calls at this time on a 1:1 basis.

Reply:
If that's the case, could you change this to loop through all results and return an array? Makes the impl simpler and it's easier and less surprising for the caller.